### PR TITLE
Implement perform sar lambda

### PR DESCRIPTION
**Pull request by laurenemms**
This is the last PR for the Formstack SAR work.

There's a bit of refactoring in it but the main goal of this PR is to create the functionality that performs the actual SAR and ties the the process of updating and retrieving from Dynamo, collecting additional data from Formstack, and writing to S3. 

I've also created some tests and a stub lambda for running in CODE.

Tested locally and will test in CODE tomorrow while this is in review.

### CODE REVIEW

File: formstack-baton-requests/project/plugins.sbt [hunk 1]
```diff
@@ -1,3 +1,4 @@
 addSbtPlugin("com.gu" % "sbt-riffraff-artifact" % "1.0.0")
 
-addSbtPlugin("net.virtual-void" % "sbt-dependency-graph" % "0.10.0-RC1")
\ No newline at end of file
+addSbtPlugin("net.virtual-void" % "sbt-dependency-graph" % "0.10.0-RC1")
```
Comment (ravimbhatt) [inline formstack-baton-requests/project/plugins.sbt]: do we need this?

File: formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/FormstackSarService.scala [hunk 2]
```diff
@@ -77,4 +86,58 @@ object FormstackSarService extends FormstackSar with LazyLogging {
     else
       decode[FormstackFormSubmissionsResponse](response.body)
   }
+
+  private def submissionsById(
```
Comment (ravimbhatt) [inline formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/FormstackSarService.scala]: this method is called submissionsById but returns FormstackSubmissionResponse. Naming is hard, i think we should try and give it a better name.

File: formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/FormstackSarService.scala [hunk 3]
```diff
@@ -77,4 +86,58 @@ object FormstackSarService extends FormstackSar with LazyLogging {
...
+
+      if(!response.is2xx) {
+        logger.error(response.body)
+      }
+
+      decode[FormstackSubmissionResponse](response.body)
+    }
+  }
+
+  private def retrieveSubmissionLabels(
```
Comment (ravimbhatt) [inline formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/FormstackSarService.scala]: same as above. may be getSubmissionQuestionAnswers?

File: formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/FormstackSarService.scala [hunk 4]
```diff
@@ -6,10 +6,12 @@ import io.circe.parser.decode
...
 import io.circe.generic.JsonCodec
+import cats.implicits._
 
 trait FormstackSar {
   def accountFormsForGivenPage(page: Int, accountToken: FormstackAccountToken): Either[Throwable, FormstackFormsResponse]
   def formSubmissionsForGivenPage(page: Int, formId: String, minTime: SubmissionTableUpdateDate, encryptionPassword: String, accountToken: FormstackAccountToken): Either[Throwable, FormstackFormSubmissionsResponse]
+  def submissionData(submissionIdEmails: List[SubmissionIdEmail], config: PerformSarLambdaConfig): Either[Throwable, List[FormstackSubmissionQuestionAnswer]]
 }
 /* Codecs for decoding accountFormsForGivenPage response */
 @JsonCodec case class FormstackForm(id: String)
```
Comment (ravimbhatt) [inline formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/FormstackSarService.scala]: would it make it easy if we removed Formstack from all these case classes? They belong to Formstak which is claer from the class and package name,

File: formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/aws/S3Client.scala [hunk 5]
```diff
@@ -13,9 +14,13 @@ case class S3CompletedPathFound(resultLocations: List[String]) extends S3StatusR
 case class S3FailedPathFound() extends S3StatusResponse
```
Comment (ravimbhatt) [inline formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/aws/S3Client.scala]: again this class is called S3Client. I personally think that the methods should not have s3 in them. we could read it as s3Client.S3FailedPathFound or s3Client.FailedPathFound

File: formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/aws/S3Client.scala [hunk 6]
```diff
@@ -64,4 +69,43 @@ object S3 extends S3Client with LazyLogging {
...
+      (meta ::: responses).mkString("\n")
+    }.mkString("\n-------------------------\n")
+  }
+
+  override def writeSuccessResult(
+    initiationId: String,
+    results: List[FormstackSubmissionQuestionAnswer],
+    config: PerformSarLambdaConfig): Either[Throwable, S3WriteSuccess] = {
+    val resultsPath = s"${config.resultsPath}/$initiationId/completed"
+    if (results.nonEmpty) {
```
Comment (ravimbhatt) [inline formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/aws/S3Client.scala]: i think we could replace this block with: 


```
 val filePath, contents =  if (results.nonEmpty) {
       logger.info("Writing SAR result to s3.")
       val filePath = s"$resultsPath/formstackSarResponse"
      val contents = formatResults(results)
      filePath, contents 
    } else {
      logger.info(s"No results found for request $initiationId. Creating NoResultsFoundForUser object.")
      val filePath = s"$resultsPath/noResultsFoundForUser"
      filePath, ""
    }

      writeToS3(config.resultsBucket, filePath, contents)
```

or better still if formatResults return empty string on the list being empty. so if condition check now return just the filepath and we could use writeToS3 as 
` 
writeToS3(config.resultsBucket, filePath, formatResults(contents))
 `

File: formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/FormstackSarService.scala (hunk 2 above)
Comment (laurenemms) [reply formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/FormstackSarService.scala]: agree - done

File: formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/FormstackSarService.scala (hunk 4 above)
Comment (laurenemms) [reply formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/FormstackSarService.scala]: agree - done

File: formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/aws/S3Client.scala (hunk 5 above)
Comment (laurenemms) [reply formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/aws/S3Client.scala]: agree - done

File: formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/FormstackSarService.scala (hunk 3 above)
Comment (laurenemms) [reply formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/FormstackSarService.scala]: done!

File: formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/aws/S3Client.scala (hunk 6 above)
Comment (laurenemms) [reply formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/aws/S3Client.scala]: Ah nice - thank you.

**Review by ravimbhatt — APPROVED** [review]
LGTM

**Pull request merged**